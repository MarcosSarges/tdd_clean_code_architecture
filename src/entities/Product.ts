export default class Product {
  private description: string;
  private price: number;

  constructor(description: string, price: number) {
    this.description = description;
    this.price = price;
  }

  getDetails() {
    return {
      description: this.description,
      price: this.price,
    };
  }
}
