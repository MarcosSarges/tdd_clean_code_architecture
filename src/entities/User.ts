import cpfValidator from './../utils/cpf/cpfValidator';

export default class User {
  private cpf: string;

  constructor(cpf: string) {
    if (!cpfValidator(cpf)) throw new Error('CPF INVALIDO');

    this.cpf = cpf;
  }

  get userInfos() {
    return {
      cpf: this.cpf,
    };
  }
}
