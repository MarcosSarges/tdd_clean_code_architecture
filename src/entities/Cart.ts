import Product from './Product';

interface Item {
  product: Product;
  quantity: number;
}

export default class Cart {
  private items: Item[] = [];

  private calculatorTotalCart() {
    return this.items.reduce(
      (prev, current) =>
        (prev += current.product.getDetails().price * current.quantity),
      0
    );
  }

  addProduct(product: Product, quantity: number) {
    this.items.push({ product, quantity });
  }

  getCartDetails() {
    return {
      items: this.items,
      totalPrice: this.calculatorTotalCart(),
    };
  }
}
