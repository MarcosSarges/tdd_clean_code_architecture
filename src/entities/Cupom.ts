import Cart from './Cart';

export default class Cupom {
  private discount: number = 0;

  constructor(code: string) {
    if (code === 'OFF50') {
      this.discount = 0.5;
    }
  }

  get discountPercentage() {
    return this.discount;
  }

  applyDiscount(cart: Cart) {
    return cart.getCartDetails.totalPrice * (1 - this.discount);
  }
}
