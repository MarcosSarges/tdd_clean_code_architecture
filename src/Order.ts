import Cart from './Cart';
import Cupom from './Cupom';
import User from './User';

export default class Order {
  checkout(cart: Cart, user: User, cupom?: Cupom) {
    const paymentTotal = cupom
      ? cupom?.applyDiscount(cart)
      : cart.getCartDetails.totalPrice;

    return {
      user: user.userInfos,
      status: 'DONE',
      items: cart.getCartDetails.items,
      paymentTotal,
    };
  }
}
