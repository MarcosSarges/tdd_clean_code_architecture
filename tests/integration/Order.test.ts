import Cart from '../../src/entities/Cart';
import Cupom from '../../src/entities/Cupom';
import Item from '../../src/entities/Item';
import Order from '../../src/entities/Order';
import Product from '../../src/entities/Product';
import User from '../../src/entities/User';

describe('Order', () => {
  let order: Order;

  beforeEach(() => {
    order = new Order();
  });

  it('Não deve fazer um pedido com cpf inválido', () => {
    const cpf = '553.566.310-55';

    expect(() => {
      const user = new User(cpf);
      const cart = new Cart();
      order.checkout(cart, user);
    }).toThrowError('CPF INVALID');
  });
  it('Deve fazer um pedido com cpf válido', () => {
    const user = new User('553.566.310-73');
    const cart = new Cart();
    const product = new Product('Short azul', 1000);
    const item1 = new Item(product, 1);

    cart.addItemInCart(item1);
    expect(order.checkout(cart, user)).toMatchSnapshot();
  });

  it('Deve fazer um pedido com 3 itens (com descrição, preço e quantidade) ', () => {
    const user = new User('553.566.310-73');
    const cart = new Cart();
    const product1 = new Product('Short azul', 1000);
    const product2 = new Product('Short rosa', 1000);
    const product3 = new Product('Short vermelho', 1000);

    const item1 = new Item(product1, 2);
    const item2 = new Item(product2, 2);
    const item3 = new Item(product3, 2);
    cart.addItemInCart(item1);
    cart.addItemInCart(item2);
    cart.addItemInCart(item3);

    expect(order.checkout(cart, user).status === 'DONE').toMatchSnapshot();
  });
  it('Deve fazer um pedido com cupom de desconto (percentual sobre o total do pedido)', () => {
    const user = new User('553.566.310-73');
    const cart = new Cart();
    const cupom = new Cupom('OFF50');
    const product1 = new Product('Short azul', 1000);
    const product2 = new Product('Short rosa', 1000);
    const product3 = new Product('Short vermelho', 1000);

    const item1 = new Item(product1, 2);
    const item2 = new Item(product2, 2);
    const item3 = new Item(product3, 2);
    cart.addItemInCart(item1);
    cart.addItemInCart(item2);
    cart.addItemInCart(item3);

    expect(order.checkout(cart, user, cupom).paymentTotal).toEqual(3000);
  });

  it('Não deve aplicar cupom de desconto expirado', () => {
    const user = new User('553.566.310-73');
    const cart = new Cart();
    // const cupom = new Cupom('OFF50');
    const product = new Product('Short azul', 1000);
    const item = new Item(product, 1);
    cart.addItemInCart(item);
    expect(order.checkout(cart, user).paymentTotal).toEqual(1000);
  });
  it.todo(
    'Deve calcular o valor do frete com base nas dimensões (altura, largura e profundidade em cm) e o peso dos produtos (em kg)'
  );
  it.todo(
    'Deve retornar o preço mínimo de frete caso ele seja superior ao valor calculado'
  );
});
