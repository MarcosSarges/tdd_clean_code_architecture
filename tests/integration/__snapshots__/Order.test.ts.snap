// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Deve fazer um pedido com 3 itens (com descrição, preço e quantidade)  1`] = `true`;

exports[`Order Deve fazer um pedido com cpf válido 1`] = `
Object {
  "items": Array [
    Item {
      "product": Product {
        "description": "Short azul",
        "height": 1,
        "length": 3,
        "price": 1000,
        "weight": 4,
        "width": 2,
      },
      "quantity": 1,
    },
  ],
  "paymentTotal": 1000,
  "status": "DONE",
  "user": Object {
    "cpf": "553.566.310-73",
  },
}
`;
